### PR TITLE
Fix concurrent run

### LINKED ISSUE #49: Warnings and crashes for FlagserPersistence with n_jobs>1
#### Description
With `FlagserPersistence` (from giotto-tda), for `n_jobs>1` , I get many notifications
```
"Error deleting flagser output file: No such file or directory"
```
and eventually a
```
TerminatedWorkerError: A worker process managed by the executor was unexpectedly terminated. This could be caused by a segmentation fault while calling the function or by an excessive memory usage causing the Operating System to kill the worker.
The exit codes of the workers are {UNKNOWN(255)}
```
It seems to be related to a memory issue, but I could not trace an increased usage (to my surprise, actually).

#### Steps/Code to Reproduce
```
import numpy as np

from gtda.point_clouds import ConsecutiveRescaling
from gtda.homology import FlagserPersistence

pc = np.random.randn(1000, 60, 2)
CR = ConsecutiveRescaling(factor=1.)
distance_matrices = CR.fit_transform(pc)
FlagserPersistence(n_jobs=9).fit_transform(distance_matrices)
```

#### Expected Results
An array of persistence diagrams.

#### Actual Results
```
---------------------------------------------------------------------------
TerminatedWorkerError                     Traceback (most recent call last)
<ipython-input-4-ee0d8999551c> in <module>
      2 CR = ConsecutiveRescaling(factor=1.)
      3 distance_matrices = CR.fit_transform(pc)
----> 4 FlagserPersistence(n_jobs=9).fit_transform(distance_matrices)

~/Libs/giotto-tda/gtda/utils/_docs.py in fit_transform_wrapper(*args, **kwargs)
    104         @wraps(original_fit_transform)
    105         def fit_transform_wrapper(*args, **kwargs):
--> 106             return original_fit_transform(*args, **kwargs)
    107         fit_transform_wrapper.__doc__ = \
    108             make_fit_transform_docs(fit_docs, transform_docs)

~/snap/miniconda3/envs/ts_rp/lib/python3.8/site-packages/sklearn/base.py in fit_transform(self, X, y, **fit_params)
    688         if y is None:
    689             # fit method of arity 1 (unsupervised transformation)
--> 690             return self.fit(X, **fit_params).transform(X)
    691         else:
    692             # fit method of arity 2 (supervised transformation)

~/Libs/giotto-tda/gtda/homology/simplicial.py in transform(self, X, y)
   1306         X = check_point_clouds(X, accept_sparse=True, distance_matrices=True)
   1307 
-> 1308         Xt = Parallel(n_jobs=self.n_jobs)(
   1309             delayed(self._flagser_diagram)(x) for x in X)
   1310 

~/snap/miniconda3/envs/ts_rp/lib/python3.8/site-packages/joblib/parallel.py in __call__(self, iterable)
   1059 
   1060             with self._backend.retrieval_context():
-> 1061                 self.retrieve()
   1062             # Make sure that we get a last message telling us we are done
   1063             elapsed_time = time.time() - self._start_time

~/snap/miniconda3/envs/ts_rp/lib/python3.8/site-packages/joblib/parallel.py in retrieve(self)
    938             try:
    939                 if getattr(self._backend, 'supports_timeout', False):
--> 940                     self._output.extend(job.get(timeout=self.timeout))
    941                 else:
    942                     self._output.extend(job.get())

~/snap/miniconda3/envs/ts_rp/lib/python3.8/site-packages/joblib/_parallel_backends.py in wrap_future_result(future, timeout)
    540         AsyncResults.get from multiprocessing."""
    541         try:
--> 542             return future.result(timeout=timeout)
    543         except CfTimeoutError as e:
    544             raise TimeoutError from e

~/snap/miniconda3/envs/ts_rp/lib/python3.8/concurrent/futures/_base.py in result(self, timeout)
    437                 raise CancelledError()
    438             elif self._state == FINISHED:
--> 439                 return self.__get_result()
    440             else:
    441                 raise TimeoutError()

~/snap/miniconda3/envs/ts_rp/lib/python3.8/concurrent/futures/_base.py in __get_result(self)
    386     def __get_result(self):
    387         if self._exception:
--> 388             raise self._exception
    389         else:
    390             return self._result

~/snap/miniconda3/envs/ts_rp/lib/python3.8/site-packages/joblib/externals/loky/_base.py in _invoke_callbacks(self)
    623         for callback in self._done_callbacks:
    624             try:
--> 625                 callback(self)
    626             except BaseException:
    627                 LOGGER.exception('exception calling callback for %r', self)

~/snap/miniconda3/envs/ts_rp/lib/python3.8/site-packages/joblib/parallel.py in __call__(self, out)
    364         with self.parallel._lock:
    365             if self.parallel._original_iterator is not None:
--> 366                 self.parallel.dispatch_next()
    367 
    368 

~/snap/miniconda3/envs/ts_rp/lib/python3.8/site-packages/joblib/parallel.py in dispatch_next(self)
    797 
    798         """
--> 799         if not self.dispatch_one_batch(self._original_iterator):
    800             self._iterating = False
    801             self._original_iterator = None

~/snap/miniconda3/envs/ts_rp/lib/python3.8/site-packages/joblib/parallel.py in dispatch_one_batch(self, iterator)
    864                 return False
    865             else:
--> 866                 self._dispatch(tasks)
    867                 return True
    868 

~/snap/miniconda3/envs/ts_rp/lib/python3.8/site-packages/joblib/parallel.py in _dispatch(self, batch)
    782         with self._lock:
    783             job_idx = len(self._jobs)
--> 784             job = self._backend.apply_async(batch, callback=cb)
    785             # A job can complete so quickly than its callback is
    786             # called before we get here, causing self._jobs to

~/snap/miniconda3/envs/ts_rp/lib/python3.8/site-packages/joblib/_parallel_backends.py in apply_async(self, func, callback)
    529     def apply_async(self, func, callback=None):
    530         """Schedule a func to be run"""
--> 531         future = self._workers.submit(SafeFunction(func))
    532         future.get = functools.partial(self.wrap_future_result, future)
    533         if callback is not None:

~/snap/miniconda3/envs/ts_rp/lib/python3.8/site-packages/joblib/externals/loky/reusable_executor.py in submit(self, fn, *args, **kwargs)
    175     def submit(self, fn, *args, **kwargs):
    176         with self._submit_resize_lock:
--> 177             return super(_ReusablePoolExecutor, self).submit(
    178                 fn, *args, **kwargs)
    179 

~/snap/miniconda3/envs/ts_rp/lib/python3.8/site-packages/joblib/externals/loky/process_executor.py in submit(self, fn, *args, **kwargs)
   1100         with self._flags.shutdown_lock:
   1101             if self._flags.broken is not None:
-> 1102                 raise self._flags.broken
   1103             if self._flags.shutdown:
   1104                 raise ShutdownExecutorError(

TerminatedWorkerError: A worker process managed by the executor was unexpectedly terminated. This could be caused by a segmentation fault while calling the function or by an excessive memory usage causing the Operating System to kill the worker.

The exit codes of the workers are {UNKNOWN(255), UNKNOWN(255)}
```
#### Versions

Linux-5.0.0-1070-oem-osp1-x86_64-with-glibc2.10
Python 3.8.5 (default, Sep  4 2020, 07:30:14) 
[GCC 7.3.0]
pyflagser 0.4.1

### PR DESCRIPTION
**Pull request by reds-heig**
#### Reference Issues/PRs

Fixes #49. 

#### What does this implement/fix? Explain your changes.

This PR use a temporary name for the result file instead of a fixed one. This prevent errors when running multiples Flagser instances in parallel.

The Bindings now throw exceptions instead of calling exit function.

#### Any other comments?

Added regression test.

### CODE REVIEW

File: pyflagser/tests/test_flagser.py
Comment (ulupo) [inline pyflagser/tests/test_flagser.py:231]: Why do we need this? It seems to just be a renaming of `flagser_unweighted`

File: pyflagser/tests/test_flagser.py
Comment (ulupo) [inline pyflagser/tests/test_flagser.py:235]: Great that you added this without introducing extra dependencies!

File: src/flagser_bindings.cpp
Comment (ulupo) [inline src/flagser_bindings.cpp:128]: "generated"?

**Review by ulupo — COMMENTED** [review]
Great!

File: pyflagser/tests/test_flagser.py
Comment (MonkeyBreaker) [reply pyflagser/tests/test_flagser.py:231]: Because my brain sometimes misses the point ...
We don't need this function, and even the function name is wrong, god dammit
Let me fix this

File: src/flagser_bindings.cpp
Comment (MonkeyBreaker) [reply src/flagser_bindings.cpp:128]: The comment is an old one, but needs a fix, let me also add it. Nice spotted.

**Review by ulupo — APPROVED** [review]
LGTM! Let's wait for the CI before merging.

**Comment by MonkeyBreaker** [comment]
@ulupo seems that we have some issues with `python3.9`.

On Mac, an memory allocation error occurs with the biggest dataset.
On Windows a problem with `Coveragy.py` occurred, do you know why ?

**Comment by ulupo** [comment]
@MonkeyBreaker I honestly don't understand why some pipelines are failing. Note that in the latest run there is nothing specific to Python 3.9, it's only about coverage files being corrupt. But it's not happening in other jobs, so I'm not sure at all what to do!

**Comment by MonkeyBreaker** [comment]
@ulupo seems that the same issue does not appear with `giotto-tda` pipelines.
Do we have different versions for `pytest` and everything related to the coverage ?

**Comment by MonkeyBreaker** [comment]
> @ulupo seems that the same issue does not appear with `giotto-tda` pipelines.
> Do we have different versions for `pytest` and everything related to the coverage ?

Well in `giotto-tda` `python39` support is waiting on `pyflagser` in order to be merged. 

I'll wait on the current run to finish and look into more details why it fails. If only `coverage` fails, maybe we could allow errors in the pipeline (only for the coverage) in order to make this PR pass.

**Comment by ulupo** [comment]
@MonkeyBreaker fixed! It appears the problem was caused by the new test because there is delicate interplay between `multiprocessing` and `pytest-cov`, see https://pytest-cov.readthedocs.io/en/latest/subprocess-support.html?highlight=multiprocessing#if-you-use-multiprocessing-pool. I added
```
pool.close()
pool.join()
```
at the end of the test and that did the trick! I don't really understand why, but OK.

I leave it to you to restore the other tests before merging.

By the way, I noticed there are some deprecation warnings coming from `flagser_bindings.cpp` which should be easy to remove, see the logs at https://dev.azure.com/maintainers/f3825528-0a61-43d7-87c2-e296dd6abb14/_apis/build/builds/3289/logs/68

**Comment by MonkeyBreaker** [comment]
@ulupo Nice catch ! What an amateurish mistake I made ...

Sure, let me enable again all the test, by the way, shall we enable again the coverage on Windows and Linux ?

**Comment by ulupo** [comment]
> Shall we enable again the coverage on Windows and Linux ?

I think it's OK to leave only the macOS tests. We run the same tests everywhere anyway.

**Comment by MonkeyBreaker** [comment]
Extensive test enable again.
Concerning the warning, there's isn't an easy cross-platform fix, I made `flagser` a trivial output generator (it does not write any file).
Now we don't need to create or remove any files !

**Comment by ulupo** [comment]
>  I made flagser a trivial output generator (it does not write any file).

This sounds great! I was always unsure why this was done, seems to add loads of I/O overhead that could slow things down a lot? Why was it thought to be needed? Why was it so easy to remove this?

**Comment by MonkeyBreaker** [comment]
> This sounds great! I was always unsure why this was done, seems to add loads of I/O overhead that could slow things down a lot?

I don't think this would slow things down by a lot, but for sure it added lots of I/O interactions.

> Why was it thought to be needed? Why was it so easy to remove it?

When I used, `flagser` I always choose to save the results into a file, and I didn't know that a output-format that could disable storing results into a file existed. It was looking into the code that I discovered this.

**Comment by MonkeyBreaker** [comment]
@ulupo Should we disable extensive test on Windows ?
Seems like it always much much slower than on other platforms.

**Comment by ulupo** [comment]
@MonkeyBreaker I don't think we should disable tests in Windows. After all, we want to check that everything works everywhere. For sure we should (in another PR IMO) invest some time in speeding these tests up. Azure gives us 60min per job and we sometimes exceed them!

This time, I will re-trigger until things work.

**Comment by MonkeyBreaker** [comment]
@ulupo maybe because of the added test, the windows pipeline will not end ?

**Comment by ulupo** [comment]
@MonkeyBreaker oh dear, they keep taking longer than 1 h. I have an idea: from now on we can either test the wheels or we test the dev install, but never both. We can merge this PR with the failures and I'll create another on CI immediately.

**Comment by MonkeyBreaker** [comment]
@ulupo I'm fine with with doing that on the short term, but on the long-term, I would prefer not to touch too much the pipeline. I'll try to analyse what's going on with the Windows run and see if some optimization is disable or if some code slow downs everything.

But at least for the moment, we can merge this PR !

Thank you again for helping into debugging the issues :)

File: pyflagser/tests/conftest.py
Comment (ulupo) [inline pyflagser/tests/conftest.py:8]: Shall we remove this line?

Comment (MonkeyBreaker) [reply pyflagser/tests/conftest.py:8]: We can, I let it because if in the future we would like to have 2 types of test suite (quick, extensive), we already know which dataset are the extensive ones. But we can remove it.

Comment (ulupo) [reply pyflagser/tests/conftest.py:8]: I'd suggest tp remove it not to confuse anyone who hasn't been following the discussions in this thread, but it's not important.

**Comment by ulupo** [comment]
> I'll try to analyse what's going on with the Windows run and see if some optimization is disable or if some code slow downs everything.

This would be very useful, of course.

**Pull request merged**
